### ISSUE TITLE
Plan-V4D: A task graph engine with support for 2D/3D graphics as well as GUI in conjuction with computer vision.

### ISSUE THREAD

**Issue by kallaballa** [opencv/opencv#22923]
The name changed from V4D to Plan-V4D because it consists of two components now.

[Tutorials](https://github.com/kallaballa/Plan-V4D/tree/next_release/modules/v4d/tutorials)

#### How it Compares

Plan-V4D is like GStreamer, but with a compile-time graph that is type-safe C++. It has the pipeline-thinking of a node-based compositor like Blender or Nuke, but implemented in code. It delivers the performance of a low-level tasking library like Intel TBB, but is specialized for the computer vision and graphics domain.

#### Release
 I have released an alpha version of the [module](https://github.com/kallaballa/Plan-V4D). See the README!

At the moment it depends on pull requests that haven't been merged yet but could work without, just slower: #22780 and #22704 

#### Currently supported
* it uses C++20
* It supports OpenGL 3.0 and ES 3.2
* Depth buffer support
* it supports usage of multiple OpenGL-contexts (using context-sharing) so multiple independent state-machines are available. Very handy when you want to integrate an external API based on OpenGL or you simply don't want/need to manage GL-states meticulously between different rendering sub-systems.
* It supports NanoVG (2D and font-rendering) out-of-the-box by using the multi-context-support.
* It supports an experimental fork of BGFX  (3D engine which i altered so it will adhere to the concurrency model of Plan-V4D)
* It supports making GUI-overlays using IMGui
* it is based on a template impementation of a task graph engine which is facilitated for parallelization
* it is based on T-API
* it supports fast videoio and manipulation of videos
* it facilitates hardware acceleration where it can (e.g. VAAPI) and makes use of interop implementations to keep frames on the GPU (in some samples all the time)
* in optimal cases it uses my 4.x fork: https://github.com/kallaballa/opencv but it is compatible with main-line
* Most of the PRs pertaining the fork have already been intergrated into 4.x and 5.0
* External dependencies: OpenCL (optionally), OpenGL, GLEW and GLFW
* Internal dependencies: opencv_core opencv_imgproc opencv_videoio opencv_video

What do you think?

**Comment by kallaballa** [comment]
https://user-images.githubusercontent.com/287266/206014208-e55e1f75-1446-42ba-9e3b-579511d27fdd.mp4

A screen capture of the full example as it looks now.

**Comment by kallaballa** [comment]
A [WebAssembly/WebGL application](https://rgl.epfl.ch/tekari?url=%2F%2Frgl.s3.eu-central-1.amazonaws.com%2Fmedia%2Fuploads%2Fwjakob%2F2018%2F08%2F27%2Firidescent-paper.txt&log=1) based on nanogui

**Comment by kallaballa** [comment]
I redesigned it into a simple object model.

https://viel-zu.org/opencv/apidoc/

Notes:
* A lot of things still not use OpenCV types and ways
* To retain NanoVG functionality I consider creating a OpenCV-wrapper for it in its own namespace. Similar to ```cv::ogl```
* I still have to write a demo without hardware acceleration
* The member functions ```makeVAcapture()```, ```makeVAwriter()```, ```captureVA()``` and ```writeVA()``` are only required till #22780 is resolved.

Is see large potential to go beyond the highgui API but I guess one of the next things to do is to implement my version of ```namedWindow``` and ```imshow```.

**Comment by kallaballa** [comment]
After reading some more I think this fits much better into cv::viz.

**Comment by kallaballa** [comment]
I put some more experiments into it to realize what this could be and I think it could be [Viz2D](https://viel-zu.org/opencv/apidoc/classkb_1_1Viz2D.html). I already called it that because if i remove the opengl context creation and windowing code then this could be a nice addition to Viz3D. Actually, once Viz3D is finished (or maybe it already works with the PR?) it could quickly become useful because it exposes the internal frame buffer as a ```cv::ogl::Texture2D```. That also removes the strict requirement (but still nice to have) for #22704 because the implementation doesn't concern about OpenGL context creation and such. Instead of removing GL/Windowing i could have a separate initialization routine that provides that optional.

**Comment by kallaballa** [comment]
I have a first version of a [wrapper for NanoVG](https://viel-zu.org/opencv/apidoc/namespacekb_1_1nvg.html). There are still some things to do (like using cv::String everywhere and cv::Mat for the transform functions) but it basically works. all the functions inside namespace kb::nvg should only be used inside [```kb::Viz2D::nanovg```](https://viel-zu.org/opencv/apidoc/classkb_1_1Viz2D.html#a16f177c4bb08b3c67badd19c4ee67fe9).
Still a lot of testing needed.

**Comment by mshabunin** [comment]
I'm wondering why is it necessary to wrap nanovg? Isn't it easier to use this library directly and maybe provide small compatibility header from OpenCV side?

_highgui_ module purpose is to provide cross-platform interface for very basic GUI applications, probably because there were no appropriate existing libraries at the time of its creation.

_viz_ module has been moved to the _opencv_contrib_ (#13069) because it was considered that this module is not useful for majority of users and sometimes adds build problems due to its dependency on VTK.

Also there is [ovis](https://github.com/opencv/opencv_contrib/tree/4.x/modules/ovis) module wrapping [Ogre3d](https://www.ogre3d.org/) library (cross-platform, HW-accelerated).

**Comment by kallaballa** [comment]
The wrapping of nanovg might be questionable but I just found it natural. I am trying to get some guidance on the whole idea. :)
You don't think it is worth it?

**Comment by kallaballa** [comment]
> The wrapping of nanovg might be questionable but I just found it natural. I am trying to get some guidance on the whole idea. :) You don't think it is worth it?

Anyway, they are a few good reasons to wrap NanoVG. NanoVG requires OpenGL which involves a complex setup (especially cross-platform). By wrapping it that initialization could be required (e.g. by exception) or even performed on use. My motivation behind wrapping NanoVG is to bring hardware accelerated 2D Vector graphics to the opencv user in an easy manner. That's also why the wrapper (mostly) provides an interface using OpenCV types and behaviour (e.g. colorspace conversions and the definition of hue).

At the moment the use of NanoVG is bound to a ```kb::viz::Viz2```object which ensures everything is setup right.
Btw. what was the motivation behind the ```cv::ogl``` wrapper?

**Comment by kallaballa** [comment]
https://user-images.githubusercontent.com/287266/208972149-c6615ae0-470a-4d6c-886a-9c6d43f9174a.mp4

New features: Multiple light- and heavy-weight windows, zooming, feeding multiples windows from one frame buffer and a couple of little things.

https://github.com/kallaballa/GCV/blob/main/src/optflow/optflow-demo.cpp

**Comment by kallaballa** [comment]
> > The wrapping of nanovg might be questionable but I just found it natural. I am trying to get some guidance on the whole idea. :) You don't think it is worth it?
> 
> Anyway, they are a few good reasons to wrap NanoVG. NanoVG requires OpenGL which involves a complex setup (especially cross-platform). By wrapping it that initialization could be required (e.g. by exception) or even performed on use. My motivation behind wrapping NanoVG is to bring hardware accelerated 2D Vector graphics to the opencv user in an easy manner. That's also why the wrapper (mostly) provides an interface using OpenCV types and behaviour (e.g. colorspace conversions and the definition of hue).
> 
> At the moment the use of NanoVG is bound to a `kb::viz::Viz2`object which ensures everything is setup right and CL/GL interop can be used in conjunction with it seamlessly. Btw. what was the motivation behind the `cv::ogl` wrapper?

Another important point is resilience against API changes in NanoVG

**Comment by kallaballa** [comment]
updated the [skeleton api docs](https://viel-zu.org/opencv/apidoc/)

**Comment by kallaballa** [comment]
Anyway there is still much to do but I am hesitant about which direction this project should go. Pushing it to contrib at start and re-evaluating based on user acceptance later would be fine for me. I believe this may become very useful to the user and then it doesn't matter if it's in contrib for starters. also if it would stay for various reasons (like dependencies) in contrib that's also an acceptable outcome to me. As long as the user has a decent way of getting hands on that code.
I actually thought about documenting it by adapting existing OpenCV tutorials to it one by one. And there is of course the topic of the minimal source commenting at the moment. :)

If OpenCV decides it doesn't want the module at all I would still continue developing it but the project would probably take quite another direction.

Would be nice to gain more clarity on that.

**Comment by kallaballa** [comment]
I ported the Viz2D library and one of the simpler demos (no input video required) to WebAssembly

https://user-images.githubusercontent.com/287266/209198199-2a4ee58c-8513-4fa6-bb44-5cdd55cc04ae.mp4

**Comment by kallaballa** [comment]
A more complete WASM build of the [font-demo](https://github.com/kallaballa/GCV/blob/main/src/font/font-demo.cpp). 
- High-DPI support works even in the browser.
- Coding as well as using the GUI feels good.
- No more html5 frontends! (except for special cases :))

Next I'll see how much more performance I can squeeze out of it, but I don't think It'll be much.

**Comment by kallaballa** [comment]
I am working right now on a wasm port of the optflow demo. I think because of no decoding and encoding overhead (I get raw video from html5 and I won't write a video for starters), it might run decently at maybe even 720p15+? Maybe more if I tweak visual settings.

**Comment by kallaballa** [comment]
An early version of the wasm port of the optflow-demo: https://viel-zu.org/opencv/optflow/index.php
I use Firefox 107.0.1 for testing. You have to press the "start capture" button to activate camera input.

**Comment by kallaballa** [comment]
Now the wasm build should work for most modern browsers.

**Comment by kallaballa** [comment]
Improved the font-demo and put it online: https://viel-zu.org/opencv/font/

**Comment by mshabunin** [comment]
> Pushing it to contrib at start and re-evaluating based on user acceptance later would be fine for me.

Yes, I think it would be fine.

**Comment by kallaballa** [comment]
alright. will take some time.,. in the meantime do you have an example for an excellent contrib-module to use as a template?

**Comment by kallaballa** [comment]
anyway, i finished two additional demos:

## Shader Demo

https://user-images.githubusercontent.com/287266/222886064-a95c96f2-d0e1-424d-9da6-6b3b68bd56df.mp4


## Beauty Demo

https://user-images.githubusercontent.com/287266/222894043-5a00d631-b4e1-4e7d-8297-cb23d33d1ed0.mp4

## I ported both to WASM
https://viel-zu.org/opencv/shader/
https://viel-zu.org/opencv/beauty/

**Comment by kallaballa** [comment]
Ported the beautification demo to WASM and added a GUI:
https://viel-zu.org/opencv/beauty/

**Comment by kallaballa** [comment]
I finished the last demo: Pedestrian detection.

https://user-images.githubusercontent.com/287266/222980241-d631f7e5-e7a3-446e-937e-bce34e194bd1.mp4

**Comment by kallaballa** [comment]
Renamed the project to [Viz2D](https://github.com/kallaballa/Viz2D) and included instructions for Ubuntu 22.04.2 LTS.

**Comment by mshabunin** [comment]
> do you have an example for an excellent contrib-module to use as a template?

Perhaps [ovis](https://github.com/opencv/opencv_contrib/tree/4.x/modules/ovis) is similar to your module. General guidance is here: https://github.com/opencv/opencv/wiki/Coding_Style_Guide and documentation writing guide is here: https://docs.opencv.org/4.x/d4/db1/tutorial_documentation.html#tutorial_documentation_quick_start

**Comment by kallaballa** [comment]
thx!

**Comment by kallaballa** [comment]
Progress:
* Transparent hardware acceleration: CL-GL sharing and VAAPI are automatically detected and therefor are now optional
* The only hardware capability required is either OpenGL 4 or OpenGL ES 3.0 (probably can be lowered to OpenGL 3 and ES 2.0)
* Improved WebAssembly support

**Comment by kallaballa** [comment]
>  https://github.com/opencv/opencv/wiki/Coding_Style_Guide
Mentions C++11. is this still a requirement? I read the C++11 will be dropped soon. Also, aren't there modules that use C++20? I am asking because at the moment Viz2D is using C++20.

**Comment by kallaballa** [comment]
From now on my efforts to port Viz2D to a OpenCV-module reside in a seperate branch: https://github.com/kallaballa/Viz2D/tree/opencv-module

**Comment by kallaballa** [comment]
> From now on my efforts to port Viz2D to a OpenCV-module reside in a seperate branch: https://github.com/kallaballa/Viz2D/tree/opencv-module

In that branch i already changed the namespace to cv::viz and changed the license to Apache2. I also wrote source documentation for all public functions: https://viel-zu.org/opencv/apidoc/namespaces.html

**Comment by kallaballa** [comment]
I finished converting the project into an OpenCV module, so i merged back to [main](https://github.com/kallaballa/Viz2D) and updated the build instructions in the README. The build requires C++20 at the moment.
That said, there is still a lot to do.

**Comment by mshabunin** [comment]
> This guide mentions C++11. Is this still a hard requirement? 

Yes, we will stick to C++11 in 4.x branch at least. For 5.x min standard version might be raised to C++14 or C++17, not sure regarding C++20.

Perhaps you can check C++ version in your module cmake and disable itself if compiler does not support C++20. Similar check has been implemented in the _cvv_ module: https://github.com/opencv/opencv_contrib/blob/6ebb38db6d0504682d72d168c4bba13a6b70895c/modules/cvv/CMakeLists.txt#L1-L4

**Comment by kallaballa** [comment]
Thx! I'll check how much effort it would be to port to C++11.

**Comment by kallaballa** [comment]
I heavily extended documentation (for the moment) in the [README](https://github.com/kallaballa/Viz2D/blob/main/README.md).
I intend to write tutorials based on the [samples](https://github.com/kallaballa/Viz2D/tree/main/modules/viz2d/samples/cpp)

**Comment by kallaballa** [comment]
> Thx! I'll check how much effort it would be to port to C++11.

NanoGUI needs C++20 so for the moment i will stick to it and check for the C++ version as suggested.

**Comment by kallaballa** [comment]
There is another thing. Since Viz2D supports OpenGL the name doesn't fit. How about renaming it to cv::viz::Vizual?

**Comment by kallaballa** [comment]
Ported from OpenGL 4.6 Compatibility to OpenGL 3.0 Core.

**Comment by kallaballa** [comment]
Implemented (transparent) parallel capturing/writing for a considerable performance boost.

**Comment by kallaballa** [comment]
> > This guide mentions C++11. Is this still a hard requirement?
> 
> Yes, we will stick to C++11 in 4.x branch at least. For 5.x min standard version might be raised to C++14 or C++17, not sure regarding C++20.
> 
> Perhaps you can check C++ version in your module cmake and disable itself if compiler does not support C++20. Similar check has been implemented in the _cvv_ module: https://github.com/opencv/opencv_contrib/blob/6ebb38db6d0504682d72d168c4bba13a6b70895c/modules/cvv/CMakeLists.txt#L1-L4

I think i found a good solution that only turns on C++20 for Viz2D while other code compiles as C++11.

First i check for C++20 support and disable the module if it isn't found:
```cmake
get_property(known_features GLOBAL PROPERTY CMAKE_CXX_KNOWN_FEATURES)
list (FIND known_features "cxx_std_20" idx)
if (${idx} LESS 0)
  message(STATUS "Module opencv_viz2d disabled because it requires C++20")
  ocv_module_disable(viz2d)
endif()
```

If C++20 is available I enable it using ```target_compile_features``` which turns on all required flags for a feature.
```cmake
target_compile_features(opencv_viz2d PRIVATE cxx_std_20)
```
No need to build all of OpenCV as C++20.

**Comment by kallaballa** [comment]
I'm not sure that's the way to do things but for now I included nanogui (which includes nanovg) as a submodule and adjust the [build instructions](https://github.com/kallaballa/Viz2D/blob/main/README.md#instructions-for-ubuntu-22042-lts)

**Comment by kallaballa** [comment]
Viz2D only depends on core modules while the samples depend on additional contrib modules. How and when should I enable building of samples and introduce the dependencies?

**Comment by kallaballa** [comment]
Also at which point should i request a review? It's not perfect but it is a working OpenCV module. Could need some guidance.

**Comment by kallaballa** [comment]
I integrated https://github.com/bloomen/cxxpool which is a MIT header-only implementation. Is that alright?

**Comment by kallaballa** [comment]
Added a new tutorial about custom video sources and sinks: https://viel-zu.org/opencv/doxygen/html/dc/d8a/viz2d_custom_source_and_sink.html

**Comment by kallaballa** [comment]
I renamed the project to V4D and create a PR draft: https://github.com/opencv/opencv_contrib/pull/3471

**Comment by kallaballa** [comment]
What kind of tests should i write? performance?

**Comment by mshabunin** [comment]
> Viz2D only depends on core modules while the samples depend on additional contrib modules. How and when should I enable building of samples and introduce the dependencies?

Something like this: https://github.com/opencv/opencv_contrib/blob/298fc7be958c10be8a8e807cf9a0dde512c13af5/modules/ovis/CMakeLists.txt#L27

> I integrated https://github.com/bloomen/cxxpool which is a MIT header-only implementation. Is that alright?

OpenCV already have it's own thread pool (`cv::parallel_for`), it is better to reuse it if possible. Adding more threads can result in conflicts, oversubscription and performance degradation.

> What kind of tests should i write? performance?

Some basic accuracy tests verifying module functionality. E.g. creating windows, showing images, etc.. Performance tests can be added in case if you expect some functions to have optimized implementation or for comparing between various HW/SW configurations.

**Comment by kallaballa** [comment]
> > Viz2D only depends on core modules while the samples depend on additional contrib modules. How and when should I enable building of samples and introduce the dependencies?
> 
> Something like this: https://github.com/opencv/opencv_contrib/blob/298fc7be958c10be8a8e807cf9a0dde512c13af5/modules/ovis/CMakeLists.txt#L27
> 
> > I integrated https://github.com/bloomen/cxxpool which is a MIT header-only implementation. Is that alright?
> 
> OpenCV already have it's own thread pool (`cv::parallel_for`), it is better to reuse it if possible. Adding more threads can result in conflicts, oversubscription and performance degradation.
> 
> > What kind of tests should i write? performance?
> 
> Some basic accuracy tests verifying module functionality. E.g. creating windows, showing images, etc.. Performance tests can be added in case if you expect some functions to have optimized implementation or for comparing between various HW/SW configurations.

Thx!

**Comment by kallaballa** [comment]
I made a rather fundamental architectural change: every context (fb, gl, nvg, ..) has it's own offscreen window and copies back and forth to the main (on-screen) frambuffer using interop. I can't observe any performance hit but i suspect some latency (which is absolutely no priority).
But the real interesting part is when it comes to other graphics APIs than OpenGL... for example, that way i could implement a make-shift (but fast!) vulkan<->opengl<->opencl interop bridge.

Would look something like that:
```c++
v4d->vk([](cv::Size sz) {
//do vulkan things
});
v4d->fb([](cv::UMat& fb){
//access the result of the vulkan context using interop bridge.
});
```
Same might work with Metal but i haven't really looked into it yet.

**Comment by kallaballa** [comment]
Implemented multi-context support for GLFW+emscripten. So for the meantime WebAssembly support only with my [fork](https://github.com/kallaballa/emscripten) till the [PR](https://github.com/emscripten-core/emscripten/pull/19233) is through.

**Comment by kallaballa** [comment]
I embedded the WASM builds of all examples and demos (except beauty-demo) into the doxygen. There is still a lot to do but it's fun having live applications for the tutorial code. :)

* The examples: https://viel-zu.org/opencv/doxygen/html/d5/dae/v4d_display_image_pipeline.html
* The demos: https://viel-zu.org/opencv/doxygen/html/d5/d87/v4d_cube.html

**Comment by kallaballa** [comment]
I am still investigating ways to share context data (like [virtual-webgl](https://github.com/greggman/virtual-webgl)) which would make things much easier but i have another solution if performance is really needed: In WASM mode i could turn off context separation entirely and advice the user to either use only one API (e.g. nanovg) or to do manual GL state-management. That would still not yield performance comparable to a desktop build but it should be **much** faster and portable (if you turn off context separation on desktop too).

Anyway I'm also doing a bit of research into the future of Graphics APIs and I think should consider WebGPU support soon (rather than vulkan). That said, I couldn't find any resources on the possibility of WebGPU-OpenGL interop - which would make things really easy.

**Comment by kallaballa** [comment]
I found a workaround based on virtual-webgl and the [minimal opengl example](https://viel-zu.org/opencv/doxygen/html/d2/deb/v4d_render_opengl.html) runs more than 3x faster :).

**Comment by kallaballa** [comment]
I implemented a (still crude and a bit buggy) camera capturing interface that has very little impact on the overall FPS.
That makes most or even all (depending on your machine) examples and demos actually usable.
Except from an input event handling api and a lot of polishing and docs, this was the last big task to a beta version :)

**Comment by kallaballa** [comment]
I replaced the [CPU-copy based](https://github.com/kallaballa/V4D/blob/5915da21a1de3d1ae57fa85a500f05120f50ba70/modules/v4d/src/util.cpp#L389-L403) camera capture code with a [GPU-copy based](https://github.com/kallaballa/V4D/blob/5915da21a1de3d1ae57fa85a500f05120f50ba70/modules/v4d/src/util.cpp#L330-L387) one (using [texImage2D](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texImage2D) on a html video element) for a little performance boost - because in WASM we have to down/up-load at some point (at least most of the times) to be able to do something with the data.

**Comment by kallaballa** [comment]
I found time to rewrite part of the webgl specific implementation. I used to rely on virtual-webgl2 to provide inter-context resource sharing but it proved to be not equipped for more complex cases. so I instead opted to copy the color buffer off a canvas (of one context) using textImage2D  (into another context) - which is a GPU copy. Addtionally i needed to blend textures using shaders.

The demos are still a bit glitchy but they work, and i finally added [beauty-demo](https://viel-zu.org/opencv/doxygen/html/d3/ddf/v4d_beauty.html)

**Comment by kallaballa** [comment]
Fixed the flicker at the cost of FPS (for webkit).

**Comment by kallaballa** [comment]
Fixed some glitches and optimized performance. e.g. cube-demo went from about 130 FPS to around 200 FPS.

Just for future reference:
**CPU**: Intel i7-1160G7
**GPU**: Intel Tiger Lake-UP4 GT2 [Iris Xe Graphics] (rev 01)
**Kernel**: 6.3.6-1-default
**Display Server**: X.Org X Server 1.21.1.8
**Browser**: Version 114.0.5735.106 (openSUSE Build) stable (64-bit)

**Comment by kallaballa** [comment]
I got **OpenGL 3.2 Core**, **OpenGL ES 3.0** and **WebGL2** working at the same time for all examples and demos. Additionally for the first time I made a broader browser test (only up-to-date browsers on different platforms). **WebKit is the fastest** and works equally well across browsers and operating systems. On Linux I found performance **problems with Firefox** that are due to an inefficiency in their teximage2d call for html5-canvases. I believe i can work around it, but I am going to file a bug anyway. **Safari doesn't work**... Anyway, the big suprise for me was **performance on mobile devices** - it's unexpectedly good (e.g. on my Xiaomi Redmi Note 8 Pro).

**Comment by kallaballa** [comment]
Updated the [documentation](https://viel-zu.org/opencv/doxygen/html/d7/dfc/v4d.html) (especially the native build instructions).

**Comment by kallaballa** [comment]
I think i solved all the big challenges and I am closing in on feature completeness (at least for a first release). Except from fixing bugs the only major thing missing are a simple callback-based input-event-system and a complete and up-to-date api documentation. 
Anyway, I think I am **finished with the tutorials and demos**.  They consist of a short description, a WASM live-demo and commented source code. 

The [tutorials](https://viel-zu.org/opencv/doxygen/html/d5/dae/v4d_display_image_pipeline.html) and The [demos](https://viel-zu.org/opencv/doxygen/html/d5/d87/v4d_cube.html) are designed to be read one after the other to give you a good overview over the key concepts of V4D. 
The demos show how to use V4D to the fullest and create programs that run mostly (the part the matters) on the GPU (when driver capabilities allow it). They are also a good starting point for your own applications because they touch many key aspects and algorithms of OpenCV.

Feedback is very welcome!

**Comment by kallaballa** [comment]
Btw. because I was interested in compiling the demos, right from the docs in the browser, I had a look at [emception](https://jprendes.github.io/emception/). Really interesting but also very slow... to slow.
Anyway I keep an eye on it. Wouldn't it be cool to be able to edit and compile c++ examples in the docs, right in place?

**Comment by kallaballa** [comment]
I made an **alpha** [release](https://github.com/kallaballa/V4D/releases/tag/0.0.5-alpha)
Debian-packages, installer and binary tarball.

have fun :)

**Comment by kallaballa** [comment]
I took time to familiarize myself with the future of extensions/technologies/APIs pertaining V4D and I think the best route to get a cross platform solution would be to put V4D in **GLES3** mode and let it run atop [**ANGLE**](https://github.com/google/angle). GLES3 is also the requirement for [**Dawn**](https://dawn.googlesource.com/dawn) which would provide a **WebGPU** implentation. WebGPU interop is implemented as a [canvas-copy operation](https://github.com/gpuweb/gpuweb/issues/18#issuecomment-911845614) (```texImage2d``` on a HTMLCanvasElement without a CPU round-trip). The same could work for [**WEBNN interop**](https://github.com/gpuweb/gpuweb/issues/2500).
Also i found [cl_intel_unified_shared_memory](https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_unified_shared_memory.html)

**Comment by kallaballa** [comment]
I half-way integrated Dawn for WebGPU support, just to realize it's not there yet for my use-case - but interop works.

**Comment by kallaballa** [comment]
I implemented a new demo that shows how to use multiple independent OpenGL contexts/states for rendering: https://viel-zu.org/opencv/doxygen/html/d4/d4a/v4d_many_cubes.html

Also, I have a new machine to test on with a 13th Gen Intel CPU and a NVIDIA 4070 Ti GPU. I started to test and optimize for those platforms too and especially NVIDIA support should be much better.

**Comment by kallaballa** [comment]
I had quite my share of problems with NanoGUI. NanoVG alone works really well. So I am gonna take a step back and replace NanoGUI/NanoVG with bgfx/imgui which supports rendering via NanoVG as well.

**Comment by kallaballa** [comment]
Won't work. Had to go through the source code to find out that custom backbuffers aren't supported in the GL-renderer.
So... NanoVG/imgui is the next try.

**Comment by kallaballa** [comment]
> Won't work. Had to go through the source code to find out that custom backbuffers aren't supported in the GL-renderer. So... NanoVG/imgui is the next try.

That's a much nicer mix of APIs!
![screenshot of a visualization with a gui](https://github.com/opencv/opencv/assets/287266/8d2d7151-d121-4394-8524-c6a6a19794e5)
I have the basics working :)

**Comment by kallaballa** [comment]
A preview of the new GUI. Originally in 1080p.

https://github.com/opencv/opencv/assets/287266/3eff9258-7f93-47ba-8bcb-975678eefcbb

**Comment by kallaballa** [comment]
I found time to fix up the web demos. they are stable and (relatively) fast now (still slower on firefox).
Anyway, more interesting I found my native-binary tests with multiple contexts. I took the [Many_Cubes-Demo](https://viel-zu.org/opencv/doxygen/html/d4/d4a/v4d_many_cubes.html) and tested performance with different numbers of contexts.

1 Cube -> ~300FPS
10 Cubes -> ~150FPS
100 Cubes -> ~ 130FPS
1000 Cubes -> ~ 12FPS

Please note that each of the cubes belongs to an independent GL-context/window.

**Comment by kallaballa** [comment]
I added a new example to the tutorial series: [example_v4d_display_image_nvg](https://viel-zu.org/opencv/doxygen/html/dc/d59/v4d_display_image_nvg.html)
It uses NanoVG to load, transform and blit an image.

Also I'm done porting all [examples](https://viel-zu.org/opencv/doxygen/html/d5/dae/v4d_display_image_pipeline.html) and [demos](https://viel-zu.org/opencv/doxygen/html/d5/d87/v4d_cube.html) and invested some time in performance optimization and polishing. **Now is a really good time to check them out!**

**Comment by kallaballa** [comment]
I created a mobile friendly design for the documentation based on [doxygen-bootstrapped](https://github.com/Velron/doxygen-bootstrapped).

https://viel-zu.org/opencv/doxygen/html/d7/dfc/v4d.html

Enjoy!

**Comment by kallaballa** [comment]
I made a new alpha-release: https://github.com/kallaballa/V4D/releases/tag/0.0.6-alpha

**Comment by kallaballa** [comment]
I implemented partial context allocation, which means you can choose which subsystems will be allocated... if you don't need NanoVG don't allocate it....
Annnnnd... I implemented a simple form of pipelining. It is fascinating how different hardware & work-load react to the pipelining.  For some demos it yields great gain while for others it is even slightly detrimental.
You can add workers now by specifying their number in the ```V4D::run``` call.

**Comment by kallaballa** [comment]
My recent experiments with pipelining lead me to the decision that at want to support easy interop with G-API in three steps.
1. Implement a Source and Sink for G-API streams
2. Implement a V4D-G-API context where you can define/pass your graph and have it interact with other V4D-contexts 
3. Use G-API as a execution backend for all V4D tasks.

Point 1 and 2 are pretty easy and I've already prototyped it. Point 3 is tricky but I've made some promising experiments and start to understand how to do it properly. However, for Point 3 I only want to support linear graph creation and I won't tackle it before release 1.0.

**Comment by kallaballa** [comment]
I investigated platform support through ANGLE and it doesn't look good. No CLGL, problems with multi-threading and more. _If_ I'm going to implement it, it will just be an optional fallback.

Also... I ported all examples & demos to have primitive pipelining support. Not all of them benefit from it, so it is turned off by default.... But the beefy ones like beauty-demo benefit greatly!

**Comment by kallaballa** [comment]
Btw. pipelining currently doesn't work for WebAssembly builds and I doubt I would be beneficial, because the WebAssembly builds spend the majority of time in the main thread to be able to access browser resources (WebGL2, video capture, etc.)

**Comment by kallaballa** [comment]
You can find the source for the many_cubes-demo [here](https://github.com/kallaballa/V4D/blob/4.x/modules/v4d/samples/many_cubes-demo.cpp). But beware, much has changed API wise, but more about that in my next post.

**Comment by kallaballa** [comment]
OK. Here's about the new API. The idea behind it is to enable **parallel/graph execution** of context calls (which run invocables) in a safe and fast manner without changing the overall API too much and with only little or no boiler plate compared to the raw algorithm. This is done by extending ```class Plan``` (which represents a serialized graph definition) So instead of executing the call right away, we store the invocable (see first rule) and **references** to the arguments. Storing references instead of values is important for automatic inference.

The five foremost API-rules:
1. All code to be inferred must be executed through a context call (except in the setup and teardown phase which may do calls without)
2. The invocable (which is passed to all context calls) must be a stateless lambda. That means you can't capture anything, not even by value, except for implicit captures.
3. Limit implicit captures to constants and thread_locals. If you must access a non-thread-local resource makes sure it is thread-safe.
4. Since we only store references to the arguments the lifetime of the arguments must be coupled to the lifetime of the Plan
5. As previously mentioned the lambda may take arguments. Carefully evaluate constness of arguments because const arguments are treated as read (in-edge) and non-const are treated as READ-WRITE (in/out-edge).

Some of these and other rules are already compile time checked (e.g. stateless lamba) and some are merely realized by convention. The compile time checks are currently implemented based on enable_if, static_assert, etc.. but should be implemented using c++ contraints and concepts)

A commented example:

```c++
#include <opencv2/v4d/v4d.hpp>
#include <opencv2/imgcodecs.hpp>

using namespace cv;
using namespace cv::v4d;

//We need to extend class Plan which has several virtual functions we could override, but in our case we only need 'setup' and 'infer'
class DisplayImageFB : public Plan {
    //Used as arguments in context calls. by being a member of the Plan object lifetime is automatically bound to it.
    UMat image_;
    UMat converted_;
public:
    //setup-phase: inferred and executed only once.
    void setup(cv::Ptr<V4D> win) override {
        //a parallel context can run in parallel in different threads opposed to a single context which is globally mutexed.
        //except for that parallel doesn't do any special context setup.
        //As you can see some lambda arguments are defined as const to mark them as in-edge in the graph.
        win->parallel([](cv::UMat& image, cv::UMat& converted, const cv::Size& sz){
            //Loads an image as a UMat (just in case we have hardware acceleration available)
            image = imread(samples::findFile("lena.jpg")).getUMat(ACCESS_READ);
            //We have to manually resize to framebuffer size
            resize(image, converted, sz);
            cvtColor(converted, converted, COLOR_RGB2BGRA);
        }, image_, converted_, win->fbSize()); // this is where the references are passed. Again, everything passed here must have a lifetime longer than the execution of the plan
    }

    //infer-phase: inferred and executed in a loop
    void infer(Ptr<V4D> win) override {
        //Create a fb context and copy the prepared image to the framebuffer.
        win->fb([](UMat& framebuffer, const cv::UMat& c){
            c.copyTo(framebuffer);
        }, converted_);
    }
};

int main() {
    //Creates a V4D object
    Ptr<V4D> window = V4D::make(960, 960, "Display an Image through direct FB access");
    //Runs the plan using in total 1 thread (= 0 extra threads).
    window->run<DisplayImageFB>(0);
}
```

**Comment by kallaballa** [comment]
For now you have to browse the other samples to see all context calls in action. a relative simple sample of how to do conditional branching in the serialized graph is [font-demo](https://github.com/kallaballa/V4D/blob/4.x/modules/v4d/samples/font-demo.cpp)

**Comment by kallaballa** [comment]
Pipelining performance demo. In this video i demonstrate how adding pipeline workers enhances performance.
First run: 1 Thread
Second run: 2 Threads
Thread run: 3 Threads

[source code](https://github.com/kallaballa/V4D/blob/4.x/modules/v4d/samples/beauty-demo.cpp)

https://github.com/opencv/opencv/assets/287266/54f69e33-fb48-4bae-bd33-d6d848feb429

**Comment by kallaballa** [comment]
And what if i want to combine multiple demos into one large graph?

https://github.com/opencv/opencv/assets/287266/0fffc39d-7b90-4bf1-b68d-cc20af2fed0d

The code (which isn't very elegant yet):

```c++
int v4d_cube_main();
int v4d_many_cubes_main();
int v4d_video_main(int argc, char **argv);
int v4d_nanovg_main(int argc, char **argv);
#define main v4d_cube_main
#include "cube-demo.cpp"
#undef main
#define main v4d_many_cubes_main
#include "many_cubes-demo.cpp"
#undef main
#define main v4d_video_main
#include "video-demo.cpp"
#undef main
#define main v4d_nanovg_main
#include "nanovg-demo.cpp"
#undef main

class MontageDemoPlan : public Plan {
	CubeDemoPlan cubePlan_;
	ManyCubesDemoPlan manyCubesPlan_;
	VideoDemoPlan videoPlan_;
	NanoVGDemoPlan nanovgPlan_;

	cv::UMat cube_;
	cv::UMat many_cubes_;
	cv::UMat video_;
	cv::UMat nanovg_;
public:
	virtual void setup(cv::Ptr<V4D> window) override {
		cubePlan_.width_ = 1920;
		cubePlan_.height_ = 1080;
		manyCubesPlan_.width_ = 1920;
		manyCubesPlan_.height_ = 1080;
		videoPlan_.width_ = 1920;
		videoPlan_.height_ = 1080;
		nanovgPlan_.width_ = 1920;
		nanovgPlan_.height_ = 1080;

		cubePlan_.setup(window);
		manyCubesPlan_.setup(window);
		videoPlan_.setup(window);
		nanovgPlan_.setup(window);
	}

	virtual void infer(cv::Ptr<V4D> window) override {
		cubePlan_.infer(window);
		window->fb([](const cv::UMat& framebuffer, cv::UMat& cube){
			framebuffer.copyTo(cube);
		}, cube_);

		manyCubesPlan_.infer(window);
		window->fb([](const cv::UMat& framebuffer, cv::UMat& many_cubes){
			framebuffer.copyTo(many_cubes);
		}, many_cubes_);

		videoPlan_.infer(window);
		window->fb([](cv::UMat& framebuffer, cv::UMat& video){
			framebuffer.copyTo(video);
		}, video_);

		nanovgPlan_.infer(window);
		window->fb([](cv::UMat& framebuffer, cv::UMat& cube, cv::UMat& many_cubes, cv::UMat& video, cv::UMat& nanovg){
			framebuffer.copyTo(nanovg);
			cv::resize(cube, framebuffer(cv::Rect(0, 0, 960, 540)), cv::Size(960, 540));
			cv::resize(many_cubes, framebuffer(cv::Rect(960, 0, 960, 540)), cv::Size(960, 540));
			cv::resize(video, framebuffer(cv::Rect(0, 540, 960, 540)), cv::Size(960, 540));
			cv::resize(nanovg, framebuffer(cv::Rect(960, 540, 960, 540)), cv::Size(960, 540));
		}, cube_, many_cubes_, video_, nanovg_);

	}

	virtual void teardown(cv::Ptr<V4D> window) override {
		cubePlan_.teardown(window);
		manyCubesPlan_.teardown(window);
		videoPlan_.teardown(window);
	}
};

int main(int argc, char** argv) {
#ifndef __EMSCRIPTEN__
	if (argc != 2) {
        cerr << "Usage: montage-demo <video-file>" << endl;
        exit(1);
    }
	constexpr double FPS = 60;
	constexpr const char* OUTPUT_FILENAME = "montage-demo.mkv";
#else
	CV_UNUSED(argc);
	CV_UNUSED(argv);
#endif
    using namespace cv::v4d;
    cv::Ptr<MontageDemoPlan> plan = new MontageDemoPlan();
    cv::Ptr<V4D> window = V4D::make(1920, 1080, "Montage Demo", ALL);
#ifndef __EMSCRIPTEN__
    //Creates a source from a file or a device
    auto src = makeCaptureSource(window, argv[1]);
    window->setSource(src);
    //Creates a writer sink (which might be hardware accelerated)
    auto sink = makeWriterSink(window, OUTPUT_FILENAME, FPS, cv::Size(1920, 1080));
    window->setSink(sink);
#endif
    window->run(plan, 0);

    return 0;
}
```

**Comment by kallaballa** [comment]
A bit of profiling and thinking made me take the following decision on choosing the ```numbers of workers``` (and later on the maximum number of parallel tasks instead). Not passing the ```number of workers``` parameter defaults to -1 which instructs V4D to choose the number automatically. At the moment that will always be one extra worker which is equal to passing 1 to ```run```, because:
1. If capture and/or write are involved, which are pretty slow atomic operations, pipelining will lead to parallel capture/write in 2 threads - just not at the same time. Therefor substantial gain is to be expected in that case.
2. If adding an extra worker doesn't yield significant gain, there is at least only the overhead of on worker.
3. This will in many cases not produce the optimal gain, but I settle for good gain, for now.

Passing 0 means no extra worker = 1 thread in total
Passing N>0 means N extra workers =  N+1 threads in total

**Comment by kallaballa** [comment]
All demos combined into one big graph doing multi-context rendering and pipelining at the same time.
There is still a lot to do, but it works :)

[source](https://github.com/kallaballa/V4D/blob/f3b370c0cea271344fc21dbb4bbd34c2f85bc8c3/modules/v4d/samples/montage-demo.cpp)

https://github.com/opencv/opencv/assets/287266/5782552a-acad-4d8c-a85b-68bfcd9d42b5

**Comment by kallaballa** [comment]
Simplified and polished a bit.

[source code](https://github.com/kallaballa/V4D/blob/09dddb1210b7f9d089097aaa7f94c1d916f66aa9/modules/v4d/samples/montage-demo.cpp)

Also made a video showing performance with and without pipelining.

https://github.com/opencv/opencv/assets/287266/241d251d-3341-49ae-83c1-42fef185691a

**Comment by kallaballa** [comment]
I wondered what combination of techniques would yield the best performance for rendering 10000 visible cubes at the same time. The short answer: minimal pipelining + multi-context rendering + rendering multiple cubes for each context yields the best result. The exact values for the number of threads and contexts matters a lot, performance wise.

My PC can render 10000 individually rotating cubes at ~220 FPS with the following code.:
```c++
#include <opencv2/v4d/v4d.hpp>

using namespace cv::v4d;
class ManyCubesDemoPlan : public Plan {
public:
	/* Demo Parameters */
	constexpr static size_t NUM_WORKERS_ = 1;
	constexpr static size_t NUM_CONTEXTS_ = 32;
	constexpr static size_t MAX_NUM_CUBES_ = 20000 / (NUM_WORKERS_ + 1);
	constexpr static std::string subtitlesFile_ = "sub.txt";

	/* OpenGL constants and variables */
	constexpr static GLuint TRIANGLES_ = 12;
	constexpr static GLuint VERTICES_INDEX_ = 0;
	constexpr static GLuint COLORS_INDEX_ = 1;

	//Cube vertices, colors and indices
	constexpr static float VERTICES_[24] = {
		// Front face
        0.5, 0.5, 0.5, -0.5, 0.5, 0.5, -0.5, -0.5, 0.5, 0.5, -0.5, 0.5,
        // Back face
        0.5, 0.5, -0.5, -0.5, 0.5, -0.5, -0.5, -0.5, -0.5, 0.5, -0.5, -0.5
	};

	constexpr static float VERTEX_COLORS_[24] = {
			0.0, 0.1, 0.7,
			0.1, 0.2, 0.7,
			0.2, 0.3, 0.7,
			0.3, 0.4, 0.7,
			0.2, 0.3, 0.7,
			0.1, 0.2, 0.7,
			0.1, 0.1, 0.7,
			0.0, 0.1, 0.7
	};

	constexpr static unsigned short TRIANGLE_INDICES_[36] = {
		// Front
        0, 1, 2, 2, 3, 0,

        // Right
        0, 3, 7, 7, 4, 0,

        // Bottom
        2, 6, 7, 7, 3, 2,

        // Left
        1, 5, 6, 6, 2, 1,

        // Back
        4, 7, 6, 6, 5, 4,

        // Top
        5, 1, 0, 0, 4, 5
	};
private:
	std::vector<GLuint> vao_;
	std::vector<GLuint> shaderProgram_;
	std::vector<GLuint> uniformTransform_;

	//Simple transform & pass-through shaders
	static GLuint load_shader() {
#if !defined(OPENCV_V4D_USE_ES3)
	    const string shaderVersion = "330";
#else
	    const string shaderVersion = "300 es";
#endif

	    const string vert =
	            "    #version " + shaderVersion
	                    + R"(
	    precision lowp float;
	    layout(location = 0) in vec3 pos;
	    layout(location = 1) in vec3 vertex_color;
	    
	    uniform mat4 transform;
	    
	    out vec3 color;
	    void main() {
	      gl_Position = transform * vec4(pos, 1.0);
	      color = vertex_color;
	    }
	)";

	    const string frag =
	            "    #version " + shaderVersion
	                    + R"(
	    precision lowp float;
	    in vec3 color;
	    
	    out vec4 frag_color;
	    
	    void main() {
	      frag_color = vec4(color, 1.0);
	    }
	)";

	    //Initialize the shaders and returns the program
	    return cv::v4d::initShader(vert.c_str(), frag.c_str(), "fragColor");
	}

	//Initializes objects, buffers, shaders and uniforms
	static void init_scene(const cv::Size& sz, GLuint& vao, GLuint& shaderProgram, GLuint& uniformTransform) {
	    glEnable (GL_DEPTH_TEST);

	    glGenVertexArrays(1, &vao);
	    glBindVertexArray(vao);

	    unsigned int triangles_ebo;
	    glGenBuffers(1, &triangles_ebo);
	    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, triangles_ebo);
	    glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof TRIANGLE_INDICES_, TRIANGLE_INDICES_,
	            GL_STATIC_DRAW);

	    unsigned int verticies_vbo;
	    glGenBuffers(1, &verticies_vbo);
	    glBindBuffer(GL_ARRAY_BUFFER, verticies_vbo);
	    glBufferData(GL_ARRAY_BUFFER, sizeof VERTICES_, VERTICES_, GL_STATIC_DRAW);

	    glVertexAttribPointer(VERTICES_INDEX_, 3, GL_FLOAT, GL_FALSE, 0, NULL);
	    glEnableVertexAttribArray(VERTICES_INDEX_);

	    unsigned int colors_vbo;
	    glGenBuffers(1, &colors_vbo);
	    glBindBuffer(GL_ARRAY_BUFFER, colors_vbo);
	    glBufferData(GL_ARRAY_BUFFER, sizeof VERTEX_COLORS_, VERTEX_COLORS_, GL_STATIC_DRAW);

	    glVertexAttribPointer(COLORS_INDEX_, 3, GL_FLOAT, GL_FALSE, 0, NULL);
	    glEnableVertexAttribArray(COLORS_INDEX_);

	    glBindVertexArray(0);
	    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
	    glBindBuffer(GL_ARRAY_BUFFER, 0);

	    shaderProgram = load_shader();
	    uniformTransform = glGetUniformLocation(shaderProgram, "transform");
	    glViewport(0,0, sz.width, sz.height);
	}

	//Renders a rotating rainbow-colored cube on a blueish background
	static void render_scene(const cv::Size& sz, const double& x, const double& y, const double& angleMod, GLuint& vao, GLuint& shaderProgram, GLuint& uniformTransform) {
	    glViewport(0,0, sz.width, sz.height);
		//Use the prepared shader program
	    glUseProgram(shaderProgram);

	    //Scale and rotate the cube depending on the current time.
	    float angle =  fmod(double(cv::getTickCount()) / double(cv::getTickFrequency()) + angleMod, 2 * M_PI);
	    double scale = 0.05;
	    cv::Matx44f scaleMat(
	            scale, 0.0, 0.0, 0.0,
	            0.0, scale, 0.0, 0.0,
	            0.0, 0.0, scale, 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f rotXMat(
	            1.0, 0.0, 0.0, 0.0,
	            0.0, cos(angle), -sin(angle), 0.0,
	            0.0, sin(angle), cos(angle), 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f rotYMat(
	            cos(angle), 0.0, sin(angle), 0.0,
	            0.0, 1.0, 0.0, 0.0,
	            -sin(angle), 0.0,cos(angle), 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f rotZMat(
	            cos(angle), -sin(angle), 0.0, 0.0,
	            sin(angle), cos(angle), 0.0, 0.0,
	            0.0, 0.0, 1.0, 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f translateMat(
	            1.0, 0.0, 0.0, 0.0,
	            0.0, 1.0, 0.0, 0.0,
	            0.0, 0.0, 1.0, 0.0,
	              x,   y, 0.0, 1.0);

	    //calculate the transform
	    cv::Matx44f transform = scaleMat * rotXMat * rotYMat * rotZMat * translateMat;
	    //set the corresponding uniform
	    glUniformMatrix4fv(uniformTransform, 1, GL_FALSE, transform.val);
	    //Bind our vertex array
	    glBindVertexArray(vao);
	    //Draw
	    glDrawElements(GL_TRIANGLES, TRIANGLES_ * 3, GL_UNSIGNED_SHORT, NULL);
	}
public:
	ManyCubesDemoPlan(cv::Size sz) : Plan(sz) {
		vao_ = std::vector<GLuint>(MAX_NUM_CUBES_);
		shaderProgram_ = std::vector<GLuint>(MAX_NUM_CUBES_);
		uniformTransform_ = std::vector<GLuint>(MAX_NUM_CUBES_);
	}

	void setup(cv::Ptr<V4D> window) override {
		for(int32_t i = 0; i < NUM_CONTEXTS_; ++i) {
			window->gl(i, [](const int32_t& ctxIdx, const cv::Size& sz, std::vector<GLuint>& vao, std::vector<GLuint>& shader, std::vector<GLuint>& uniformTrans){
				CV_UNUSED(ctxIdx);
				for(size_t j = 0; j < MAX_NUM_CUBES_ / NUM_CONTEXTS_; ++j) {
					init_scene(sz, vao[j], shader[j], uniformTrans[j]);
				}
			}, size(), vao_, shaderProgram_, uniformTransform_);
		}
	}

	void infer(cv::Ptr<V4D> window) override {
		window->gl([](){
			//Clear the background
			glClearColor(0.0, 0.0, 0.0, 1);
			glClear(GL_COLOR_BUFFER_BIT);
		});

		//Render using multiple OpenGL contexts
		for(int32_t i = 0; i < NUM_CONTEXTS_; ++i) {
			window->gl(i, [](const int32_t& ctxIdx, const cv::Size& sz, std::vector<GLuint>& vao, std::vector<GLuint>& shader, std::vector<GLuint>& uniformTrans){
				for(size_t j = 0; j < MAX_NUM_CUBES_ / NUM_CONTEXTS_; ++j) {
					size_t gw = MAX_NUM_CUBES_ / NUM_CONTEXTS_;
					size_t k = ctxIdx * NUM_CONTEXTS_ + j;
					size_t xpos =  k % gw;
					size_t ypos =  k / gw;
					double x = (xpos * (2.0 /gw)) -1;
					double y = (ypos * (2.0 /gw)) -1;
					double angle = ((x * sin((x * y) * 2 * M_PI)) + (y * cos((x * y) * 2 * M_PI)) / 2.0);
					render_scene(sz, x, y + 1, angle, vao[k], shader[k], uniformTrans[k]);
				}
			}, size(), vao_, shaderProgram_, uniformTransform_);
		}

		window->write();
	}
};

int main() {
	cv::Ptr<ManyCubesDemoPlan> plan = new ManyCubesDemoPlan(cv::Size(1920, 1080));
    cv::Ptr<V4D> window = V4D::make(plan->size(), "Many Cubes Demo", NANOVG);

	constexpr double FPS = 60;
	constexpr const char* OUTPUT_FILENAME = "many_cubes-demo.mkv";
    auto sink = makeWriterSink(window, OUTPUT_FILENAME, FPS, plan->size());
    window->setPrintFPS(true);
    window->setShowFPS(false);
    window->setShowTracking(false);
    window->setSink(sink);
    window->run(plan, ManyCubesDemoPlan::NUM_WORKERS_);

    return 0;
}
```

**Comment by seclorum** [comment]
@kallaballa Impressive bit of showcasing.  Bookmarked for further entertainment .. this thread is quite interesting!

**Comment by kallaballa** [comment]
I forgot to remove video writing. Without it i get 1380 FPS :). I have to redo my experiments.

**Comment by seclorum** [comment]
@kallaballa Will you push a WebAssembly build of things at some point in the future?

**Comment by kallaballa** [comment]
@seclorum Next alpha release!

**Comment by kallaballa** [comment]
> I forgot to remove video writing. Without it i get 1380 FPS :). I have to redo my experiments.

Code was buggy. The maximum i can get out of this now is ~13 FPS with 10000 individually rotation cubes and **many** GL-Contexts.

Code:
```c++
#include <opencv2/v4d/v4d.hpp>

using namespace cv::v4d;
class ManyCubesDemoPlan : public Plan {
public:
	/* Demo Parameters */
	constexpr static size_t NUM_WORKERS_ = 16;
	constexpr static size_t NUM_CONTEXTS_ = 150 / (NUM_WORKERS_ + 1);
	constexpr static size_t NUM_CUBES_ = 10000;

	/* OpenGL constants and variables */
	constexpr static GLuint TRIANGLES_ = 12;
	constexpr static GLuint VERTICES_INDEX_ = 0;
	constexpr static GLuint COLORS_INDEX_ = 1;

	//Cube vertices, colors and indices
	constexpr static float VERTICES_[24] = {
		// Front face
        0.5, 0.5, 0.5, -0.5, 0.5, 0.5, -0.5, -0.5, 0.5, 0.5, -0.5, 0.5,
        // Back face
        0.5, 0.5, -0.5, -0.5, 0.5, -0.5, -0.5, -0.5, -0.5, 0.5, -0.5, -0.5
	};

	constexpr static float VERTEX_COLORS_[24] = {
			0.0, 0.1, 0.7,
			0.1, 0.2, 0.7,
			0.2, 0.3, 0.7,
			0.3, 0.4, 0.7,
			0.2, 0.3, 0.7,
			0.1, 0.2, 0.7,
			0.1, 0.1, 0.7,
			0.0, 0.1, 0.7
	};

	constexpr static unsigned short TRIANGLE_INDICES_[36] = {
		// Front
        0, 1, 2, 2, 3, 0,

        // Right
        0, 3, 7, 7, 4, 0,

        // Bottom
        2, 6, 7, 7, 3, 2,

        // Left
        1, 5, 6, 6, 2, 1,

        // Back
        4, 7, 6, 6, 5, 4,

        // Top
        5, 1, 0, 0, 4, 5
	};
private:
	std::vector<GLuint> vao_;
	std::vector<GLuint> shaderProgram_;
	std::vector<GLuint> uniformTransform_;

	//Simple transform & pass-through shaders
	static GLuint load_shader() {
#if !defined(OPENCV_V4D_USE_ES3)
	    const string shaderVersion = "330";
#else
	    const string shaderVersion = "300 es";
#endif

	    const string vert =
	            "    #version " + shaderVersion
	                    + R"(
	    precision lowp float;
	    layout(location = 0) in vec3 pos;
	    layout(location = 1) in vec3 vertex_color;
	    
	    uniform mat4 transform;
	    
	    out vec3 color;
	    void main() {
	      gl_Position = transform * vec4(pos, 1.0);
	      color = vertex_color;
	    }
	)";

	    const string frag =
	            "    #version " + shaderVersion
	                    + R"(
	    precision lowp float;
	    in vec3 color;
	    
	    out vec4 frag_color;
	    
	    void main() {
	      frag_color = vec4(color, 1.0);
	    }
	)";

	    //Initialize the shaders and returns the program
	    return cv::v4d::initShader(vert.c_str(), frag.c_str(), "fragColor");
	}

	//Initializes objects, buffers, shaders and uniforms
	static void init_scene(const cv::Size& sz, GLuint& vao, GLuint& shaderProgram, GLuint& uniformTransform) {
	    glEnable (GL_DEPTH_TEST);

	    glGenVertexArrays(1, &vao);
	    glBindVertexArray(vao);

	    unsigned int triangles_ebo;
	    glGenBuffers(1, &triangles_ebo);
	    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, triangles_ebo);
	    glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof TRIANGLE_INDICES_, TRIANGLE_INDICES_,
	            GL_STATIC_DRAW);

	    unsigned int verticies_vbo;
	    glGenBuffers(1, &verticies_vbo);
	    glBindBuffer(GL_ARRAY_BUFFER, verticies_vbo);
	    glBufferData(GL_ARRAY_BUFFER, sizeof VERTICES_, VERTICES_, GL_STATIC_DRAW);

	    glVertexAttribPointer(VERTICES_INDEX_, 3, GL_FLOAT, GL_FALSE, 0, NULL);
	    glEnableVertexAttribArray(VERTICES_INDEX_);

	    unsigned int colors_vbo;
	    glGenBuffers(1, &colors_vbo);
	    glBindBuffer(GL_ARRAY_BUFFER, colors_vbo);
	    glBufferData(GL_ARRAY_BUFFER, sizeof VERTEX_COLORS_, VERTEX_COLORS_, GL_STATIC_DRAW);

	    glVertexAttribPointer(COLORS_INDEX_, 3, GL_FLOAT, GL_FALSE, 0, NULL);
	    glEnableVertexAttribArray(COLORS_INDEX_);

	    glBindVertexArray(0);
	    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
	    glBindBuffer(GL_ARRAY_BUFFER, 0);

	    shaderProgram = load_shader();
	    uniformTransform = glGetUniformLocation(shaderProgram, "transform");
	    glViewport(0,0, sz.width, sz.height);
	}

	//Renders a rotating rainbow-colored cube on a blueish background
	static void render_scene(const cv::Size& sz, const double& x, const double& y, const double& angleMod, GLuint& vao, GLuint& shaderProgram, GLuint& uniformTransform) {
	    glViewport(0,0, sz.width, sz.height);
		//Use the prepared shader program
	    glUseProgram(shaderProgram);

	    //Scale and rotate the cube depending on the current time.
	    float angle =  fmod(double(cv::getTickCount()) / double(cv::getTickFrequency()) + angleMod, 2 * M_PI);
	    double scale = 0.01;
	    cv::Matx44f scaleMat(
	            scale, 0.0, 0.0, 0.0,
	            0.0, scale, 0.0, 0.0,
	            0.0, 0.0, scale, 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f rotXMat(
	            1.0, 0.0, 0.0, 0.0,
	            0.0, cos(angle), -sin(angle), 0.0,
	            0.0, sin(angle), cos(angle), 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f rotYMat(
	            cos(angle), 0.0, sin(angle), 0.0,
	            0.0, 1.0, 0.0, 0.0,
	            -sin(angle), 0.0,cos(angle), 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f rotZMat(
	            cos(angle), -sin(angle), 0.0, 0.0,
	            sin(angle), cos(angle), 0.0, 0.0,
	            0.0, 0.0, 1.0, 0.0,
	            0.0, 0.0, 0.0, 1.0);

	    cv::Matx44f translateMat(
	            1.0, 0.0, 0.0, 0.0,
	            0.0, 1.0, 0.0, 0.0,
	            0.0, 0.0, 1.0, 0.0,
	              x,   y, 0.0, 1.0);

	    //calculate the transform
	    cv::Matx44f transform = scaleMat * rotXMat * rotYMat * rotZMat * translateMat;
	    //set the corresponding uniform
	    glUniformMatrix4fv(uniformTransform, 1, GL_FALSE, transform.val);
	    //Bind our vertex array
	    glBindVertexArray(vao);
	    //Draw
	    glDrawElements(GL_TRIANGLES, TRIANGLES_ * 3, GL_UNSIGNED_SHORT, NULL);
	}
public:
	ManyCubesDemoPlan(cv::Size sz) : Plan(sz) {
		vao_ = std::vector<GLuint>(NUM_CUBES_);
		shaderProgram_ = std::vector<GLuint>(NUM_CUBES_);
		uniformTransform_ = std::vector<GLuint>(NUM_CUBES_);
	}

	void setup(cv::Ptr<V4D> window) override {
		for(int32_t i = 0; i < NUM_CONTEXTS_; ++i) {
			window->gl(i, [](const int32_t& ctxIdx, const cv::Size& sz, std::vector<GLuint>& vao, std::vector<GLuint>& shader, std::vector<GLuint>& uniformTrans){
				for(size_t j = 0; j < NUM_CUBES_ / NUM_CONTEXTS_; ++j) {
					size_t k = ctxIdx * (double(NUM_CUBES_) / NUM_CONTEXTS_) + j;
					init_scene(sz, vao[k], shader[k], uniformTrans[k]);
				}
			}, size(), vao_, shaderProgram_, uniformTransform_);
		}
	}

	void infer(cv::Ptr<V4D> window) override {
		window->gl([](){
			//Clear the background
			glClearColor(0.0, 0.0, 0.0, 1);
			glClear(GL_COLOR_BUFFER_BIT);
		});

		//Render using multiple OpenGL contexts
		for(int32_t i = 0; i < NUM_CONTEXTS_; ++i) {
			window->gl(i, [](const int32_t& ctxIdx, const cv::Size& sz, std::vector<GLuint>& vao, std::vector<GLuint>& shader, std::vector<GLuint>& uniformTrans){
				for(size_t j = 0; j < NUM_CUBES_ / NUM_CONTEXTS_; ++j) {
					size_t gw = sqrt(NUM_CUBES_);
					size_t k = ctxIdx * (double(NUM_CUBES_) / NUM_CONTEXTS_) + j;
					size_t xpos =  k % gw;
					size_t ypos =  k / gw;
					double x = (xpos * (2.0 / gw)) -1;
					double y = (ypos * (2.0 / gw)) -1;
					double angle = ((x * sin((x * y) * 2 * M_PI)) + (y * cos((x * y) * 2 * M_PI)) / 2.0);
					render_scene(sz, x, y, angle, vao[k], shader[k], uniformTrans[k]);
				}
			}, size(), vao_, shaderProgram_, uniformTransform_);
		}
	}
};

int main() {
	cv::Ptr<ManyCubesDemoPlan> plan = new ManyCubesDemoPlan(cv::Size(1920, 1080));
    cv::Ptr<V4D> window = V4D::make(plan->size(), "Many Cubes Demo", NONE);

	constexpr double FPS = 60;
	constexpr const char* OUTPUT_FILENAME = "many_cubes-demo.mkv";
    window->setPrintFPS(true);
    window->setShowFPS(false);
    window->setShowTracking(false);
    window->run(plan, ManyCubesDemoPlan::NUM_WORKERS_);

    return 0;
}
```

**Comment by kallaballa** [comment]
10000 Cubes seems simply a bit to much. If I render 5000 cubes i get around 140 FPS with the same settings.

**Comment by kallaballa** [comment]
> @kallaballa Will you push a WebAssembly build of things at some point in the future?

Alas, I have to inform you that I completely dropped WASM support, after long consideration. The reason simply is that with current technologies (and i mainly mean WebGL) V4D will never be able to reflect in WASM builds what it can do on native platforms. That has mainly 5 reasons:
1. WebGL calls must happen in the main thread. which extremely reduces the potential for concurrency.
2. WebGL has only 8-16 contexts. V4D does multi-context rendering and that is where it shines but for a serious application you  run out of contexts very quickly
3. WebGL doesn't support cross-context sharing which requires a copy-from-canvas workaround that is (though it is a gpu to gpu copy) much slower than just sharing a texture.
4. There are no GPU/NPU implemenations of common opencv algorithms which obviously makes things even slower.
5. Supporting emscripten specific workarounds and limitations slowed down development _a lot_.

I am still very interested in a WASM build, but I think it is wise to wait for some emerging technologies (WebGPU, WEBNN, ...) and also for some developments in opencv.

Anyway, the code is now much cleaner, easier to maintain and much easier to read.

**Comment by kallaballa** [comment]
hi!
I am sorry I disappeared but I had severe health issues. I am back on it :)

**Comment by kallaballa** [comment]
I knead bgfx into submission and made it forget about its own concurrency model.

### Decode -> Draw 10,000 Shaded Cubes -> Encode

2 Threads: ~65 FPS
3 Threads: ~111 FPS
7 Threads: ~134 FPS

https://github.com/user-attachments/assets/3389da2c-8a05-42fc-b086-5e0ca0c75e76

**Comment by kallaballa** [comment]
Much has changed while the samples should do the same. I wrote an introduction to what is now called Plan-V4D (until i eventually entierly split the two components: Plan a computation graph engine and V4D a 2D/3D graphics runtime). Check it out: [Introduction](https://github.com/kallaballa/Plan-V4D/blob/4.x/README.md)

Build instructions will follow.

**Comment by kallaballa** [comment]
Among other things (like instructions for building packages) i added build instructions or Ubuntu noble (latest LTS). For other distributions I added instructions on how to use a chroot for the build. Anyway, those instructions shouldn't be hard to adapt to other distributions.

Next: Working on Mac OS X build instructions
Then: Complete the API docs
Finally: Finalize a beta release with packages.

**Comment by kallaballa** [comment]
On the side I am working on a 3D game as the most complex demo/workload yet. Rocks and mist are generated by raymarching distance fields. The whole scene is live-generated like that and goes on forever in a fractal fashion. It renders at around 110FPS@1080p on a Intel A770 GPU and a Intel i9 13900K CPU.


https://github.com/user-attachments/assets/9658a844-f784-4025-af03-21c2b387ac72

**Comment by kallaballa** [comment]
Plan-V4D on my Anbernic RG351M

https://github.com/user-attachments/assets/4cd73569-21c1-4222-81db-31de97353926

**Comment by collapsingcheek** [comment]
> Plan-V4D on my Anbernic RG351M
> 
>  ambernic2.webm





https://lnkr.fun/2788889

**Comment by kallaballa** [comment]
Plan-V4D is largely feature complete. Busy with beta-release work (docs, tutorials, cross-platform, pushing things upstream and packages).

**Comment by kallaballa** [comment]
Made techdemo of the 3D video game I am working on. It is based on Plan-V4D

https://www.youtube.com/watch?v=IxMKj1ywQpQ

**Comment by kallaballa** [comment]
Updated the [README](https://github.com/kallaballa/Plan-V4D/tree/next_release/README.md) and the [Tutorials](https://github.com/kallaballa/Plan-V4D/tree/next_release/modules/v4d/tutorials)

This should make things much clearer.
At the moment i am working on a flatpak release-demo.

**Comment by kallaballa** [comment]
I entirely rewrote the "Describe the feature and motivation" of the github issue.

**Comment by kallaballa** [comment]
A re-implementation of window_QT.*pp (imshow window using QT-style and counting ~4000 lines) using Plan-V4D (only ~700 lines). It is also much faster. gonna present some figures later.

https://github.com/user-attachments/assets/6d0581be-f9da-4cb0-8ea6-c1b39ba15444

**Comment by kallaballa** [comment]
see: #29935
